sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GoneAndRetryWithRetryPolicy.java

```diff
@@ -16,189 +15,324 @@
 import com.azure.cosmos.implementation.RetryWithException;
 import com.azure.cosmos.implementation.RxDocumentServiceRequest;
 import com.azure.cosmos.implementation.apachecommons.lang.time.StopWatch;
+import com.azure.cosmos.implementation.apachecommons.lang.tuple.Pair;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import reactor.core.publisher.Mono;
 
 import java.time.Duration;
 
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
 public class GoneAndRetryWithRetryPolicy extends RetryPolicyWithDiagnostics {
     private final static Logger logger = LoggerFactory.getLogger(GoneAndRetryWithRetryPolicy.class);
-    private final static int DEFAULT_WAIT_TIME_IN_SECONDS = 30;
-    private final static int MAXIMUM_BACKOFF_TIME_IN_SECONDS = 15;
-    private final static int INITIAL_BACKOFF_TIME = 1;
-    private final static int BACK_OFF_MULTIPLIER = 2;
-
-    private final RxDocumentServiceRequest request;
-    private volatile int attemptCount = 1;
-    private volatile int attemptCountInvalidPartition = 1;
-    private volatile int currentBackoffSeconds = GoneAndRetryWithRetryPolicy.INITIAL_BACKOFF_TIME;
-    private volatile RetryWithException lastRetryWithException;
+    private final GoneRetryPolicy goneRetryPolicy;
+    private final RetryWithRetryPolicy retryWithRetryPolicy;
     private final StopWatch durationTimer = new StopWatch();
```

Comment:
in the new code in some places this is accessed without synchronized block that may cause potential thread-safety issue.

two options:
1. ideally this should become `private volatile durationMs` to access safely from different threads. that way we wouldn't need any synchronization block around it. (this is more preferred) as volatile is more preferred than synchronized in this case.
2. access this StopWatch everywhere with synchrnozied block.

Reply:
Fixed in next iteration